tests/components/knx/test_services.py

```diff
@@ -0,0 +1,166 @@
+"""Test KNX services."""
+from homeassistant.const import STATE_OFF, STATE_ON
+from homeassistant.core import HomeAssistant
+
+from .conftest import KNXTestKit
+
+
+async def test_send(hass: HomeAssistant, knx: KNXTestKit):
+    """Test `knx.send` service."""
+    test_address = "1/2/3"
+    await knx.setup_integration({})
+
+    # send DPT 1 telegram
+    await hass.services.async_call(
+        "knx", "send", {"address": test_address, "payload": True}, blocking=True
+    )
+    await knx.assert_write(test_address, True)
+
+    # send raw DPT 5 telegram
+    await hass.services.async_call(
+        "knx", "send", {"address": test_address, "payload": [99]}, blocking=True
+    )
+    await knx.assert_write(test_address, (99,))
+
+    # send "percent" DPT 5 telegram
+    await hass.services.async_call(
+        "knx",
+        "send",
+        {"address": test_address, "payload": 99, "type": "percent"},
+        blocking=True,
+    )
+    await knx.assert_write(test_address, (0xFC,))
+
+    # send "temperature" DPT 9 telegram
+    await hass.services.async_call(
+        "knx",
+        "send",
+        {"address": test_address, "payload": 21.0, "type": "temperature"},
+        blocking=True,
+    )
+    await knx.assert_write(test_address, (0x0C, 0x1A))
+
+    # send multiple telegrams
+    await hass.services.async_call(
+        "knx",
+        "send",
+        {"address": [test_address, "2/2/2", "3/3/3"], "payload": 99, "type": "percent"},
+        blocking=True,
+    )
+    await knx.assert_write(test_address, (0xFC,))
+    await knx.assert_write("2/2/2", (0xFC,))
+    await knx.assert_write("3/3/3", (0xFC,))
+
+
+async def test_read(hass: HomeAssistant, knx: KNXTestKit):
+    """Test `knx.read` service."""
+    await knx.setup_integration({})
+
+    # send read telegram
+    await hass.services.async_call("knx", "read", {"address": "1/1/1"}, blocking=True)
+    await knx.assert_read("1/1/1")
+
+    # send multiple read telegrams
+    await hass.services.async_call(
+        "knx",
+        "read",
+        {"address": ["1/1/1", "2/2/2", "3/3/3"]},
+        blocking=True,
+    )
+    await knx.assert_read("1/1/1")
+    await knx.assert_read("2/2/2")
+    await knx.assert_read("3/3/3")
+
+
+async def test_event_register(hass: HomeAssistant, knx: KNXTestKit):
+    """Test `knx.event_register` service."""
+    events = []
+    test_address = "1/2/3"
+
+    def listener(event):
+        events.append(event)
+
+    await knx.setup_integration({})
+    hass.bus.async_listen("knx_event", listener)
```

Comment:
We have this helper:
https://github.com/home-assistant/core/blob/c1b18f48678dd40a333ea05a588be1caf5b2a050/tests/common.py#L1106